rust/src/mqtt/mqtt_message.rs

```diff
@@ -0,0 +1,181 @@
+pub struct MQTTPublishData {
+    pub topic: String,
+    pub message_id: Option<u16>,
+    pub message: Vec<u8>,
+    pub properties: Option<Vec<MQTTProperty>>, // MQTT 5.0
+}
+
+#[derive(Debug)]
+pub struct MQTTMessageIdOnly {
+    pub message_id: u16,
+    pub reason_code: Option<u8>,               // MQTT 5.0
+    pub properties: Option<Vec<MQTTProperty>>, // MQTT 5.0
+}
+
+#[derive(Debug)]
+pub struct MQTTSubscribeTopicData {
+    pub topic_name: String,
+    pub qos: u8,
+}
+
+#[derive(Debug)]
+pub struct MQTTSubscribeData {
+    pub message_id: u16,
+    pub topics: Vec<MQTTSubscribeTopicData>,
+    pub properties: Option<Vec<MQTTProperty>>, // MQTT 5.0
+}
+
+#[derive(Debug)]
+pub struct MQTTSubackData {
+    pub message_id: u16,
+    pub qoss: Vec<u8>,
+    pub properties: Option<Vec<MQTTProperty>>, // MQTT 5.0
+}
+
+#[derive(Debug)]
+pub struct MQTTUnsubscribeData {
+    pub message_id: u16,
+    pub topics: Vec<String>,
+    pub properties: Option<Vec<MQTTProperty>>, // MQTT 5.0
+}
+
+#[derive(Debug)]
+pub struct MQTTUnsubackData {
+    pub message_id: u16,
+    pub properties: Option<Vec<MQTTProperty>>, // MQTT 5.0
+    pub reason_codes: Option<Vec<u8>>,         // MQTT 5.0
+}
+
+#[derive(Debug)]
+pub struct MQTTAuthData {
+    pub reason_code: u8,                       // MQTT 5.0
+    pub properties: Option<Vec<MQTTProperty>>, // MQTT 5.0
+}
+
+#[derive(Debug)]
+pub struct MQTTDisconnectData {
+    pub reason_code: Option<u8>,               // MQTT 5.0
+    pub properties: Option<Vec<MQTTProperty>>, // MQTT 5.0
+}
+
+impl MQTTMessage {
+    pub fn message_type_string(&self) -> String {
+        match self.op {
+            crate::mqtt::mqtt_message::MQTTOperation::CONNECT(_) => "CONNECT",
+            crate::mqtt::mqtt_message::MQTTOperation::CONNACK(_) => "CONNACK",
+            crate::mqtt::mqtt_message::MQTTOperation::PUBLISH(_) => "PUBLISH",
+            crate::mqtt::mqtt_message::MQTTOperation::PUBACK(_) => "PUBACK",
+            crate::mqtt::mqtt_message::MQTTOperation::PUBREC(_) => "PUBREC",
+            crate::mqtt::mqtt_message::MQTTOperation::PUBREL(_) => "PUBREL",
+            crate::mqtt::mqtt_message::MQTTOperation::PUBCOMP(_) => "PUBCOMP",
+            crate::mqtt::mqtt_message::MQTTOperation::SUBSCRIBE(_) => "SUBSCRIBE",
+            crate::mqtt::mqtt_message::MQTTOperation::SUBACK(_) => "SUBACK",
+            crate::mqtt::mqtt_message::MQTTOperation::UNSUBSCRIBE(_) => "UNSUBSCRIBE",
+            crate::mqtt::mqtt_message::MQTTOperation::UNSUBACK(_) => "UNSUBACK",
+            crate::mqtt::mqtt_message::MQTTOperation::PINGREQ => "PINGREQ",
+            crate::mqtt::mqtt_message::MQTTOperation::PINGRESP => "PINGRESP",
+            crate::mqtt::mqtt_message::MQTTOperation::DISCONNECT(_) => "DISCONNECT",
+            crate::mqtt::mqtt_message::MQTTOperation::AUTH(_) => "AUTH",
+            _ => "UNASSIGNED",
+        }
+        .to_string()
+    }
+
+    pub fn message_type_id(&self) -> u32 {
+        match self.op {
```

Comment:
Why do we need this function ?
Do not we have this number in the header ?

Reply:
Correct. Changed the code in `detect.rs` and removed this function.